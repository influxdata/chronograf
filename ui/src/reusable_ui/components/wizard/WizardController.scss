/*
  Wizard Controller Styles
  ------------------------------------------------------------------------------
*/

.wizard-step-title {
  text-align: center;
  @include no-user-select();
}

.wizard-controller {
  display: flex;
  flex-direction: column;
  min-width: 100%;
  justify-content: space-between;
  align-items: center;
}

.wizard-skip-link {
  flex-shrink: 0;
}

.wizard-tip-text {
  min-width: 100%;
  color: $g12-forge;
  text-align: center;
  p {
    font-weight: 600;
  }
}
