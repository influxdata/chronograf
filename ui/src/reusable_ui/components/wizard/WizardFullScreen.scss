/*
  Full Screen Wizard Styles
  ------------------------------------------------------------------------------
*/

.wizard-splash {
  display: inline-flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
  position: relative;
  min-width: 100%;
  min-height: 100%;
}

.wizard-container {
  padding: 30px;
  display: inline-flex;
  border-radius: $radius;
  width: 60%;
  min-width: 800px;
  max-width: 1200px;
  background-color: $g1-raven;
}