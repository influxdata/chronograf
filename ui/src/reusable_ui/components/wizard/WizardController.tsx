// Libraries
import React, {PureComponent, ReactElement} from 'react'

// Components
import WizardProgressBar from 'src/reusable_ui/components/wizard/WizardProgressBar'

// Types
import {WizardStepProps, Step} from 'src/types/wizard'
import {StepStatus} from 'src/reusable_ui/constants/wizard'

import {ErrorHandling} from 'src/shared/decorators/errors'

interface State {
  steps: Step[]
  currentStepIndex: number
}

interface Props {
  children: Array<ReactElement<WizardStepProps>>
  handleSkip?: () => void
  skipLinkText?: string
}

@ErrorHandling
class WizardController extends PureComponent<Props, State> {
  public static defaultProps: Partial<Props> = {
    skipLinkText: 'Skip',
  }

  public static getDerivedStateFromProps(props: Props, state: State) {
    let {currentStepIndex} = state
    const {children} = props

    const childSteps = React.Children.map(
      children,
      (child: ReactElement<WizardStepProps>, i) => {
        const isComplete = child.props.isComplete()
        if (currentStepIndex === -1 && !isComplete) {
          currentStepIndex = i
        }
        return {
          title: child.props.title,
          stepStatus: isComplete ? StepStatus.Complete : StepStatus.Incomplete,
        }
      }
    )
    if (currentStepIndex === -1) {
      currentStepIndex = childSteps.length - 1
    }

    return {steps: childSteps, currentStepIndex}
  }

  constructor(props: Props) {
    super(props)
    this.state = {
      steps: [],
      currentStepIndex: -1,
    }
  }

  public render() {
    const {steps, currentStepIndex} = this.state
    const currentChild = this.CurrentChild

    return (
      <div className="wizard-controller">
        <div className="progress-header">
          <h3 className="wizard-step-title">{currentChild.props.title}</h3>
          <WizardProgressBar
            handleJump={this.jumpToStep}
            steps={steps}
            currentStepIndex={currentStepIndex}
          />
          {this.tipText}
        </div>
        {currentChild}
        {this.skipLink}
      </div>
    )
  }

  private incrementStep = () => {
    this.setState(prevState => {
      return {
        currentStepIndex: prevState.currentStepIndex + 1,
      }
    })
  }

  private decrementStep = () => {
    this.setState(prevState => {
      return {
        currentStepIndex: prevState.currentStepIndex - 1,
      }
    })
  }

  private jumpToStep = (jumpIndex: number) => () => {
    this.setState({
      currentStepIndex: jumpIndex,
    })
  }

  private get CurrentChild(): JSX.Element {
    const {children, handleSkip} = this.props
    const {currentStepIndex, steps} = this.state
    const lastStep = currentStepIndex === steps.length - 1

    const advance = lastStep ? handleSkip : this.incrementStep
    const retreat = currentStepIndex === 0 ? null : this.decrementStep

    let currentChild
    if (React.Children.count(children) === 1) {
      currentChild = children
    } else {
      currentChild = children[currentStepIndex]
    }

    return React.cloneElement<WizardStepProps>(currentChild, {
      increment: advance,
      decrement: retreat,
      lastStep,
    })
  }

  private get tipText() {
    const {currentStepIndex} = this.state
    const {children} = this.props

    let currentChild
    if (React.Children.count(children) === 1) {
      currentChild = children
    } else {
      currentChild = children[currentStepIndex]
    }

    const {
      props: {tipText},
    } = currentChild

    if (tipText) {
      return (
        <div className="wizard-tip-text">
          <p>{tipText}</p>
        </div>
      )
    }
    return null
  }

  private get skipLink() {
    const {handleSkip, skipLinkText} = this.props

    if (handleSkip) {
      return (
        <button
          className="btn btn-xs btn-primary btn-link wizard-skip-link"
          onClick={handleSkip}
        >
          {skipLinkText}
        </button>
      )
    }
    return null
  }
}

export default WizardController
