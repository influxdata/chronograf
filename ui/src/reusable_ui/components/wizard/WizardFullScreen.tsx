// Libraries
import React, {PureComponent, ReactElement} from 'react'

// Components
import WizardController from 'src/reusable_ui/components/wizard/WizardController'

import {WizardStepProps} from 'src/reusable_ui/components/wizard/WizardStep'

import {ErrorHandling} from 'src/shared/decorators/errors'

interface Props {
  children: Array<ReactElement<WizardStepProps>>
  skipLinkText?: string
  handleSkip?: () => void
  switchLinkText?: string
  handleSwitch?: () => void
  isUsingAuth: boolean
}

@ErrorHandling
class WizardFullScreen extends PureComponent<Props> {
  public render() {
    return (
      <div className="wizard-splash auth-image">
        <div className="wizard-container">{this.WizardController}</div>
        <p className="auth-credits">
          Powered by <span className="icon cubo-uniform" /> InfluxData
        </p>
      </div>
    )
  }

  private get WizardController() {
    const {
      children,
      skipLinkText,
      handleSkip,
      handleSwitch,
      switchLinkText,
      isUsingAuth,
    } = this.props

    if (children) {
      return (
        <WizardController
          handleSkip={handleSkip}
          skipLinkText={skipLinkText}
          handleSwitch={handleSwitch}
          switchLinkText={switchLinkText}
          isUsingAuth={isUsingAuth}
          jumpStep={0}
        >
          {children}
        </WizardController>
      )
    }

    return null
  }
}

export default WizardFullScreen
