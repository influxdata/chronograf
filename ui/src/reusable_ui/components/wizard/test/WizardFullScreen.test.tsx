import React from 'react'
import {shallow} from 'enzyme'

import WizardFullScreen from 'src/reusable_ui/components/wizard/WizardFullScreen'
import WizardController from 'src/reusable_ui/components/wizard/WizardController'

describe('WizardFullScreen', () => {
  let wrapper

  const wrapperSetup = (override = {}) => {
    const props = {
      children: null,
      skipLinkText: undefined,
      handleSkip: undefined,
      isUsingAuth: false,
      ...override,
    }

    return shallow(<WizardFullScreen {...props} />)
  }

  beforeEach(() => {
    jest.resetAllMocks()
    wrapper = wrapperSetup()
  })

  it('mounts without exploding', () => {
    expect(wrapper).toHaveLength(1)
  })

  it('renders no WizardController component', () => {
    expect(wrapper.find(WizardController)).toHaveLength(0)
  })

  it('matches snapshot with minimal props', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('with children', () => {
    it('renders one WizardController component', () => {
      wrapper = wrapperSetup({children: {}})
      expect(wrapper.find(WizardController)).toHaveLength(1)
    })

    it('matches snapshot with children props', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})
