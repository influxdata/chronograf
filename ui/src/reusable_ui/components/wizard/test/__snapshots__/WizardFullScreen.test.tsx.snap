// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WizardFullScreen matches snapshot with minimal props 1`] = `
<div
  className="wizard-splash auth-image"
>
  <div
    className="wizard-container"
  />
  <p
    className="auth-credits"
  >
    Powered by 
    <span
      className="icon cubo-uniform"
    />
     InfluxData
  </p>
</div>
`;

exports[`WizardFullScreen with children matches snapshot with children props 1`] = `
<div
  className="wizard-splash auth-image"
>
  <div
    className="wizard-container"
  />
  <p
    className="auth-credits"
  >
    Powered by 
    <span
      className="icon cubo-uniform"
    />
     InfluxData
  </p>
</div>
`;
